### PR TITLE
Better Initial Chat Experience

### PR DESCRIPTION
**Pull request by thebearjew**
## New Inbox Components

Adds a new button `StartNewChat` to the top of the inbox in place of `ChatFilterRows` when the user has not chats.

Additionally this adds a button at the bottom of the inbox, `BuildTeam` when the user is not part of a any teams (and thus has no chats), which takes the user to the teams page.

screen shot 2018-07-06 at 5 13 54 pm

screen shot 2018-07-09 at 9 50 37 am

## Storybook
Most of the work in this PR involves rewriting the stories for `Chat/Inbox`. Previously only the inbox rows were being rendered in a story. Now the stories will render an entire inbox including the header components (`StartNewConversation` or `ChatFilteredRow`), the row components, and `BuildTeam` component if it applies.

### CODE REVIEW

**Comment by thebearjew** [comment]
@keybase/react-hackers This could use a look

**Comment by chrisnojima** [comment]
- [ ] merge after the release

File: shared/chat/inbox/index.stories.js [hunk 1]
```diff
@@ -0,0 +1,455 @@
+// @flow
```
Comment (thebearjew) [inline shared/chat/inbox/index.stories.js]: Github isn't showing the new story file diff, just noting to make it show up.

File: shared/chat/inbox/index.desktop.js [hunk 2]
```diff
@@ -137,20 +139,25 @@ class Inbox extends PureComponent<Props, State> {
...
-            filter={this.props.filter}
-            onNewChat={this._prepareNewChat}
-            onSetFilter={this.props.onSetFilter}
-            hotkeys={isDarwin ? ['command+n', 'command+k'] : ['ctrl+n', 'ctrl+k']}
-            onHotkey={this.props.onHotkey}
-            filterFocusCount={this.props.filterFocusCount}
-            onSelectUp={this.props.onSelectUp}
-            onSelectDown={this.props.onSelectDown}
-          />
+          {this.props.showNewChat ? (
```
Comment (chrisnojima) [inline shared/chat/inbox/index.desktop.js]: I think we should change some of this plumbing. We should make a container that makes this decision, maybe ChatInboxHeader (or something), that would be connected and derive showNewChat and would then render one of these 2 components (aka inbox/index.desktop.js would just have <ChatInboxHeader filterFocusCount={} />

File: shared/chat/inbox/row/start-new-chat.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import {globalStyles, globalColors, globalMargins, isMobile} from '../../../styles'
+
+type Props = {
+  onNewChat: () => void,
+}
+
+const StartNewChat = (props: Props) => {
+  return (
+    <Box style={styleContainer}>
+      <ClickableBox style={{...stylesClickableBox}} onClick={props.onNewChat}>
```
Comment (chrisnojima) [inline shared/chat/inbox/row/start-new-chat.js]: lets use styleSheetCreate on new components
also
```
{...stylesClickableBox } will make a new object every time
```

File: shared/chat/inbox/row/start-new-chat.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+const stylesClickableBox = {
+  alignItems: 'center',
+  flexDirection: 'row',
+}
+
+const propsIconCompose = {
+  color: globalColors.blue,
+  fontSize: 16,
+  style: {},
```
Comment (chrisnojima) [inline shared/chat/inbox/row/start-new-chat.js]: 🔥

File: shared/chat/inbox/row/start-new-chat.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+const propsIconComposeMobile = {
```
Comment (chrisnojima) [inline shared/chat/inbox/row/start-new-chat.js]: I assume this is cause icon was complaining about style in flow? we have a helper for this in common-adapters just 
```
import {iconCastPlatformStyles, Icon, Box, ClickableBox, Text} from '../../../common-adapters'
```

**Comment by thebearjew** [comment]
@keybase/react-hackers Updated after review - good for another look

File: shared/chat/inbox/index.native.js [hunk 6]
```diff
@@ -183,21 +184,20 @@ class Inbox extends React.PureComponent<Props, State> {
       this.props.showSmallTeamsExpandDivider && (
         <BigTeamsDivider toggle={this.props.toggleSmallTeamsExpanded} />
       )
+    const buildTeam = this.props.showBuildATeam && <BuildTeam onBuildTeam={this.props.onBuildTeam} />
```
Comment (chrisnojima) [inline shared/chat/inbox/index.native.js]: lets connect BuiltdTeam and move onBuildTeam to itself

File: shared/chat/inbox/index.native.js [hunk 7]
```diff
@@ -183,21 +184,20 @@ class Inbox extends React.PureComponent<Props, State> {
...
+    const HeadComponent = (
+      <ChatInboxHeader
+        filterFocusCount={this.props.filterFocusCount}
+        focusFilter={this.props.focusFilter}
+        onNewChat={this.props.onNewChat}
```
Comment (chrisnojima) [inline shared/chat/inbox/index.native.js]: lets copy onNewChat to its connector

File: shared/chat/inbox/row/chat-inbox-header/index.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  isLoading: boolean,
+  onHotkey: (cmd: string) => void,
+  onNewChat: () => void,
+  onSelectDown: () => void,
+  onSelectUp: () => void,
+  onSetFilter: (filter: string) => void,
+  showNewChat: boolean,
+}
+
+const ChatInboxHeader = (props: Props) =>
```
Comment (chrisnojima) [inline shared/chat/inbox/row/chat-inbox-header/index.js]: the other option is to have 3 connectors here, one for each component, and this one which just figures out showNewChat, thats likely cleaner

**Comment by thebearjew** [comment]
@keybase/react-hackers good to go again

File: shared/chat/inbox/container/index.js [hunk 9]
```diff
@@ -14,47 +23,24 @@ const mapStateToProps = (state: TypedState, {routeState}) => {
...
     filter,
     isLoading: !state.chat2.loadingMap.isEmpty(),
     neverLoaded: state.chat2.metaMap.isEmpty(),
+    showBuildATeam,
   }
 }
 
 const mapDispatchToProps = (dispatch: Dispatch, {routeState, setRouteState, navigateAppend}) => ({
+  // Route to the teams tab and open the NewTeamDialog component
+  _onBuildTeam: () => dispatch(Route.navigateTo([teamsTab])),
```
Comment (chrisnojima) [inline shared/chat/inbox/container/index.js]: this isn't used right

File: shared/chat/inbox/index.types.js.flow [hunk 10]
```diff
@@ -33,18 +33,17 @@ export type RouteState = I.RecordOf<{
 
 export type Props = {
   children?: React$Element<any>,
+  focusFilter: () => void, // withStateHandler function
   filter: string,
   filterFocusCount: number,
   isLoading: boolean,
   nowOverride?: number, // just for dumb rendering
+  onBuildTeam: () => void, // routes to team tab
```
Comment (chrisnojima) [inline shared/chat/inbox/index.types.js.flow]: unused

File: shared/chat/inbox/container/index.js [hunk 11]
```diff
@@ -14,47 +23,24 @@ const mapStateToProps = (state: TypedState, {routeState}) => {
   const smallTeamsExpanded = routeState.get('smallTeamsExpanded')
   const rowMetadata = filter ? filteredRowData(state) : normalRowData(state, smallTeamsExpanded)
   const _selectedConversationIDKey = Constants.getSelectedConversation(state)
+  const showBuildATeam = state.teams.teamnames.size === 0
```
Comment (chrisnojima) [inline shared/chat/inbox/container/index.js]: unused

File: shared/stories/storybook.shared.js [hunk 12]
```diff
@@ -24,9 +40,21 @@ let uniqueProviderKey = 1
...
+   * GatewayDest and GatewayProvider need to be wrapped by the Provider here in
+   * order for storybook to correctly mock connected components inside of
+   * popups.
+   * React.Fragment is used to render StorybookErrorBoundary and GatewayDest as
+   * children to GatewayProvider which only takes one child
+   */
   return (story: () => React.Node) => (
     <Provider key={`provider:${uniqueProviderKey++}`} store={createStore(state => state, merged)}>
-      <StorybookErrorBoundary children={story()} />
+      <GatewayProvider key="gatewayprovider">
```
Comment (chrisnojima) [inline shared/stories/storybook.shared.js]: should double check that any popups that exist in stories still work

Comment (thebearjew) [reply shared/stories/storybook.shared.js]: @chrisnojima Just went through al the stories and tested each popup. Fixed two (`fs` and `Input/*`).

File: shared/chat/conversation/messages/set-explode-popup/index.desktop.js [hunk 13]
```diff
@@ -104,6 +104,7 @@ export default (props: Props) => {
   return (
     <FloatingMenu
       attachTo={props.attachTo}
+      position="top center"
```
Comment (thebearjew) [inline shared/chat/conversation/messages/set-explode-popup/index.desktop.js]: Suggestion from @buoyad to explicitly set the floating menu position to `top center` such that if the position doesn't calculate, the floating box won't render below the input area.

Defaults to `bottom center`: https://github.com/keybase/client/blob/master/shared/common-adapters/floating-box.desktop.js#L50

This doesn't change the position on desktop or mobile from my testing.

File: shared/chat/conversation/messages/set-explode-popup/index.desktop.js [hunk 14]
```diff
@@ -112,3 +113,5 @@ export default (props: Props) => {
     />
   )
 }
+
+export default SetExplodePopup
```
Comment (thebearjew) [inline shared/chat/conversation/messages/set-explode-popup/index.desktop.js]: Named container component in order to update the PropProvider in the story.

File: shared/chat/inbox/index.native.js (hunk 7 above)
Comment (thebearjew) [reply shared/chat/inbox/index.native.js]: Discussed offline to leave this as is for the moment.

File: shared/stories/storybook.shared.js [hunk 15]
```diff
@@ -24,9 +40,21 @@ let uniqueProviderKey = 1
...
+        <React.Fragment>
+          <StorybookErrorBoundary children={story()} />
+          <GatewayDest component={DestBox} name="popup-root" />
```
Comment (thebearjew) [inline shared/stories/storybook.shared.js]: @chrisnojima Danny mentioned that `GatewayDest` won't render for stories have popups but don't use `createPropProvider`.
We discussed it briefly and concluded that using `createPropProvider` is a reasonable requirement for rendering popups. Added a comment to `FloatingMenu` at his suggestion.

**Comment by thebearjew** [comment]
@keybase/react-hackers Could use another look por favor.

File: shared/chat/inbox/row/build-team/container.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+import {compose, connect, setDisplayName} from '../../../../util/container'
+import type {TypedState, Dispatch} from '../../../../util/container'
+import BuildTeam from '.'
+
+const mapStateToProps = (state: TypedState) => ({
+  showBuildATeam: state.teams.teamnames.size === 0,
+})
+
+const mapDispatchToProps = (dispatch: Dispatch) => ({
+  // Route to the teams tab and open the NewTeamDialog component
```
Comment (buoyad) [inline shared/chat/inbox/row/build-team/container.js]: Are you missing a key for the dialog here? this is routing to the root of the teams tab

File: shared/chat/inbox/row/index.stories.js [hunk 17]
```diff
@@ -149,6 +150,7 @@ const load = () => {
           hasUnread={true}
         />
         <BigTeamChannel {...commonChannel} teamname="Keybase" channelname="#video-games" isMuted={true} />
+
```
Comment (buoyad) [inline shared/chat/inbox/row/index.stories.js]: 🔥

File: shared/common-adapters/floating-menu.js [hunk 18]
```diff
@@ -1,4 +1,11 @@
 // @flow
+
+// For stories, all popups using FloatingMenu will need to have a PropProvider
```
Comment (buoyad) [inline shared/common-adapters/floating-menu.js]: This should go in `floating-box` (sry if I told you here heh)

File: shared/stories/storybook.shared.js (hunk 12 above)
Comment (buoyad) [inline shared/stories/storybook.shared.js]: no need for `key` anymore

File: shared/chat/inbox/row/build-team/container.js (hunk 16 above)
Comment (thebearjew) [reply shared/chat/inbox/row/build-team/container.js]: I revisited the designs and realized it should not open the newteamdialog

File: shared/chat/inbox/row/index.stories.js (hunk 17 above)
Comment (thebearjew) [reply shared/chat/inbox/row/index.stories.js]: File should have been deleted anyway

File: shared/common-adapters/floating-menu.js (hunk 18 above)
Comment (thebearjew) [reply shared/common-adapters/floating-menu.js]: Updated to both floating-box desktop and native

**Pull request merged**